### ISSUE TITLE
Allow alternative handling of classification flags.

### ISSUE THREAD

**Issue by abellgithub** [PDAL/PDAL#1701]
LAS 1.0 - 1.3 puts classification flags in the classification field.  LAS 1.4 breaks them out into their own field.  Current code creates a special field `ClassFlags` for the LAS 1.4 field.  As an option, it might be handy to have the 1.0 - 1.3 classification flags placed in the same `ClassFlags` field and clear the associated bits in the `Classification` field.  This will allow more standardized processing of LAS files of various versions.

**Comment by lc-guy** [comment]
fwiw, this is still an issue. PDAL seems to misinterpret classification values written with flags in 1.2 LAS files.
See https://github.com/laspy/laspy/issues/180 for more details.

**Comment by abellgithub** [comment]
There's no error here. PDAL simply reports classification as it is defined in the LAS specification. If you want the classification value without the flag bits, you can simply mask it off:

```
int classVal = (classification & 0x1F);
```

**Comment by craigds** [comment]
I disagree that this isn't an issue. Just stumbled across this after very much confusion why we were getting a classification of 34. I eventually figured out that PDRF 3 only supports values 0-31, and so 34 doesn't make any sense at all.

Unfortunately I can't just mask off the extra bits unless I _know_ what PDRF I'm giving to PDAL, since PDRFs 6+ can handle classifications 0-255, so 34 would be valid there.

I presume this bit masking is not required for PDRFs 6+? Since the classification flags are stored separately

I think the sane thing for PDAL to do would be to return the actual classification value with the flag bits masked away, and use the flag bits to populate the classification flags struct, same as it presumably does in PDRF 6+

**Comment by craigds** [comment]
> If you want the classification value without the flag bits, you can simply mask it off

> UPDATE: I also get the error when I convert the file to 1.4 format before adding the classification info.

PDAL itself carries bad values forward when converting PDRF 3 to PDRF 7. so I think even if the expectation is that the  user masks some bits off the classification values, PDAL itself still needs a fix for this when translating between PDRFs

**Comment by craigds** [comment]
source file: [auckland_3_2.zip](https://github.com/PDAL/PDAL/files/9770235/auckland_3_2.zip)

minimal reproducer:

```bash
# output PDRF ID and classification counts of input (PDRF 3) file, showing classification 34 is present (actually classification 2)
echo '[
{"type": "readers.las", "filename": "auckland_3_2.laz"},
{"type": "filters.stats", "dimensions": "Classification", "count": "Classification"}]' | pdal pipeline --stdin --metadata=STDOUT | jq -SC '.stages["readers.las"].dataformat_id, .stages["filters.stats"].statistic[0].counts'

# Convert LAS 1.2 to COPC (LAS 1.4) using PDAL
echo '[
{"type": "readers.las", "filename": "auckland_3_2.laz"},
{"type": "writers.copc", "filename": "auckland_3_2.copc.laz"}]' | pdal pipeline --stdin

# output PDRF ID and classification counts of output (PDRF 7) file, showing classification 34 still present
echo '[
{"type": "readers.las", "filename": "auckland_3_2.copc.laz"},
{"type": "filters.stats", "dimensions": "Classification", "count": "Classification"}]' | pdal pipeline --stdin --metadata=STDOUT | jq -SC '.stages["readers.las"].dataformat_id, .stages["filters.stats"].statistic[0].counts'
```

**Comment by abellgithub** [comment]
In your test data, the classification of 34 is valid. It represents a synthetic  point with a classification of 2. See the LAS 1.3 specification.

PDAL has no idea that you want to treat the classification field as a separate classification value with flags or as a single value. This ticket is suggesting that PDAL could create a flags field if you tell it to do so in order to facilitate this separation of meaning.

**Comment by craigds** [comment]
> In your test data, the classification of 34 is valid

that's right - in the source data (PDRF 3) the classification of 34 actually means classification *2* plus some flags.

What I am saying is that PDAL should be translating that to classification 2 plus some flags in the output (PDRF 7) file, but it is _not_ - the output file has classification 34 (plus no flags)

**Comment by abellgithub** [comment]
PDAL isn't just a LAS -> LAS translator. We provide the data to users through an API and users may well be expecting a classification value to be 34, because that's what's in the file. There are many, many examples where modifying the source value wouldn't be proper UNLESS the user requests it. This is what this issue acknowledges, but nobody has spent the time to do the necessary work yet. We would consider good code in a PR that addresses this issue.

**Comment by abellgithub** [comment]
Also, if you want to just strip away the class flags on your data, you can do this with the range filter by limiting the classifications to be in the range [0, 31].

**Comment by craigds** [comment]
> users may well be expecting a classification value to be 34, because that's what's in the file

my reading of the LAS spec says that the classification value is 2, and there are additionally some flags. The fact that they're stored in the same byte is an implementaion detail, and indeed later PDRFs moved the same flags elsewhere. PDAL reading the value as 34 is a bug in my opinion, especially when PDAL also has separate handling for flags. The PDRF 3 reader should consider the flag bits as flags, and not part of the classification value.

**Comment by abellgithub** [comment]
The LAS 1.3 spec refers to the classification field that "represents the class attributes of a point". PDAL does not have any concept of bits. Sorry.

**Comment by abellgithub** [comment]
It occurred to me that this is probably best handled with a filter.

**Comment by craigds** [comment]
I've been researching the background of this. i think i might be close to understanding why this behaviour was changed originally:

Originally PDAL did what i think the correct behaviour should be: it populated ClassFlags and changed the class value to mask out the flags.

In 2020 @jrobichaud opened an issue about it: #3092.
His complaint was that PDAL was 'corrupting' his classification value 71 and turning it into 7 (71 mod 32)

However, I think it generated a bit of confusion where he says:

> [The Point cloud used as input is a 1.4 format created using pdal "writers.las", not a 1.2 format.](https://github.com/PDAL/PDAL/issues/3092#issuecomment-633627365)

The actual LAS file he supplies at the top of the page indeed is a "LAS 1.4" file, but has

```
"dataformat_id": 3,
```
PDRF 3 in LAS 1.4 is documented as having a 5-bit value and 3-bit flags.

So then presumably `readers.las` was reading the file with `{ClassFlags = 2, Class = 7}`, and then `writers.las` was _not preserving_ the class flags: `{ClassFlags = 0, Class = 7}` when writing to PDRF 3 (the default PDRF).

That is, I think the original bug existed in `writers.las`, rather than `readers.las`. I have further described the bug (which still exists) with a reproducible example at #3878.


The fix in #3095 worked around the issue by changing the code in `readers.las` to not populate ClassFlags. This works as long as you don't try to write to PDRF 6+.

I would recommend we fix the issue in #3878 and revert the original change.
